### PR TITLE
Send DHCP requests in parallel on all interfaces to improve boot time

### PR DESCRIPTION
**Pull request by anjalidk**
We can drastically improve boot times on some machines, or BM or VM
shapes, which have more than 1 interfaces, if we send DHCP requests 
in parallel on all interfaces.
Add ip=single-dhcp as a kernel boot parameter, to send DHCP requests in
parallel on all interfaces. Thus, DHCP on each interface is overlapped with
other interfaces, reducing boot times for multiple interface machines.
This can help reduce the boot time by 1 minute per interface.
In addition, we may also save another 40 secs delay per interface, if
the link on which we try to send DHCP is in DOWN state. This can be reduced
by giving the following kernel command line parameter while booting -
rd.net.timeout.carrier=<timeout>.

Signed-off-by: Anjali Kulkarni <anjali.k.kulkarni@oracle.com>
---

This pull request changes...

## Checklist
- [x ] I have tested it locally
- [ ] I have reviewed and updated any documentation if relevant
- [ ] I am providing new code and test(s) for it

Fixes #

### CODE REVIEW

**Comment by anjalidk** [comment]
Hi - the test (15) seems to have failed due to power down of the machine? It should have no relation to my fix - is this a real issue ?

**Comment by anjalidk** [comment]
Hi - can someone pls help with this merge?

**Comment by danimo** [comment]
@tblume can you take a look at this?

**Comment by tblume** [comment]
> @tblume can you take a look at this?

I've tested the option on a iscsiroot ibft setup with 2 network interfaces and got an error:

-->
[   17.569282] dracut-cmdline[197]: ///lib/dracut/hooks/cmdline/98-parse-ip-opts.sh@68(source): for autoopt in $(str_replace "$autoconf" "," " ")
[   17.570748] dracut-cmdline[197]: ///lib/dracut/hooks/cmdline/98-parse-ip-opts.sh@69(source): case $autoopt in
[   17.571970] dracut-cmdline[197]: ///lib/dracut/hooks/cmdline/98-parse-ip-opts.sh@81(source): '[' -n 1 ']'
[   17.573109] dracut-cmdline[197]: ///lib/dracut/hooks/cmdline/98-parse-ip-opts.sh@81(source): '[' -z '' ']'
[   17.574274] dracut-cmdline[197]: ///lib/dracut/hooks/cmdline/98-parse-ip-opts.sh@82(source): die 'Sorry, '\''ip=single-dhcp'\'' does not make sense for multiple interface configurations'
[   17.576148] dracut-cmdline[197]: //lib/dracut-lib.sh@460(die): echo '<24>dracut: FATAL: Sorry, '\''ip=single-dhcp'\'' does not make sense for multiple interface configurations'
[   17.577903] dracut-cmdline[197]: //lib/dracut-lib.sh@461(die): echo '<24>dracut: Refusing to continue'
[   17.579072] dracut-cmdline[197]: //lib/dracut-lib.sh@465(die): echo 'warn dracut: FATAL: "Sorry, '\''ip=single-dhcp'\'' does not make sense for multiple interface configurations"'
[   17.580877] dracut-cmdline[197]: //lib/dracut-lib.sh@466(die): echo 'warn dracut: Refusing to continue'
--<

anjalidk does that mean that ip=single-dhcp is exclusiv to other ip= options?

**Comment by anjalidk** [comment]
Hi,

Since ip=single-dhcp means that it chooses the boot device based on 
which one gets the DHCP response first, I thought it does not make sense 
to have other ip= options with it. If there is a use case for having 
other ip= options, I can accommodate that.

However, can you send me the ip= options you gave which resulted in this 
error? That will help me understand the use case and whether single-dhcp 
is valid for that case or not.

Anjali

**Comment by anjalidk** [comment]
@tblume see my response above

**Comment by tblume** [comment]
> Hi, Since ip=single-dhcp means that it chooses the boot device based on which one gets the DHCP response first, I thought it does not make sense to have other ip= options with it. If there is a use case for having other ip= options, I can accommodate that. 

Agreed, testing with ip=dhcp leads to the same error message.
Could you please also update the dracut.cmdline manpage?

> However, can you send me the ip= options you gave which resulted in this error? That will help me understand the use case and whether single-dhcp is valid for that case or not. Anjali

Seems the ip= parameters get set implicitly at iBFT boot, they aren't visible in the logs but the command line parser sees them.
Below a log from a boot without the ip=single-dhcp parameter:

-->
Nov 24 08:05:11 localhost dracut-cmdline[201]: Using kernel command line parameters: rd.driver.pre=btrfs rd.driver.pre=scsi_dh_alua rd.driver.pre=scsi_dh_emc rd.driver.pre=scsi_dh_rdac rd.dr
iver.pre=dm_multipath rd.iscsi.ibft=1 rd.iscsi.firmware=1 root=/dev/mapper/360014057268ef2b5751436d84a0b3c0e-part2 rootfstype=btrfs rootflags=rw,relatime,space_cache,subvolid=256,subvol=/@,s
ubvol=@ BOOT_IMAGE=/boot/vmlinuz-5.9.8-2-default root=UUID=3fdc3143-8e24-49f0-b499-d8a004a1c614 splash=verbose mitigations=auto console=tty0 console=ttyS0
Nov 24 08:05:11 localhost kernel: iBFT detected.
Nov 24 08:05:11 localhost kernel: iscsi: registered transport (tcp)
Nov 24 08:05:11 localhost kernel: input: ImExPS/2 Generic Explorer Mouse as /devices/platform/i8042/serio1/input/input2
Nov 24 08:05:12 localhost dracut-cmdline[201]: Multiple ip= arguments: assuming rd.neednet=1
--<

We should probably opt to just ignore additional ip= parameters for the iBFT boot case, since the config is supposed to be provided by firmware anyway.

File: modules.d/35network-legacy/dhcp-multi.sh
Comment (tblume) [inline modules.d/35network-legacy/dhcp-multi.sh:112]: Grepping for dhclient won't work on SUSE since we are using wicked instead.

**Comment by anjalidk** [comment]
@tblume

Thanks! Since in the dhcp-multi.sh (or in ifup.sh without my changes) we 
call "dhclient", hence I am grep-ing for dhclient. Is wicked like 
network manager which needs to be explicitly enabled? I do not see any 
reference to it in ifup.sh.

Does this code in ifup.sh run on SUSE? Because the code in ifup.sh only 
calls dhclient.

Anjali

**Comment by anjalidk** [comment]
@tblume

Thanks!

On 11/23/20 23:55, tblume wrote:
>
>     Hi, Since ip=single-dhcp means that it chooses the boot device
>     based on which one gets the DHCP response first, I thought it does
>     not make sense to have other ip= options with it. If there is a
>     use case for having other ip= options, I can accommodate that.
>
> Agreed, testing with ip=dhcp leads to the same error message.
> Could you please also update the dracut.cmdline manpage?
>
ANJALI> Yes, will do!
>
>     However, can you send me the ip= options you gave which resulted
>     in this error? That will help me understand the use case and
>     whether single-dhcp is valid for that case or not. Anjali
>
> Seems the ip= parameters get set implicitly at iBFT boot, they aren't 
> visible in the logs but the command line parser sees them.
> Below a log from a boot without the ip=single-dhcp parameter:
>
ANJALI> If it is not too much trouble could you send me output of cat 
/proc/cmdline with the system booted (with ibft and hence multiple ip= 
options enabled)? If it's not easily available, then fine.

**Comment by tblume** [comment]
> @tblume Thanks! Since in the dhcp-multi.sh (or in ifup.sh without my changes) we call "dhclient", hence I am grep-ing for dhclient. Is wicked like network manager which needs to be explicitly enabled? I do not see any reference to it in ifup.sh. Does this code in ifup.sh run on SUSE? Because the code in ifup.sh only calls dhclient

You couldn't see, because:

https://github.com/dracutdevs/dracut/pull/563

hasn't been merged yet.
Still, we are using this at SUSE as default, without the need to explicitly enable it.

There is also work to support wicked as a separate module via:

https://github.com/dracutdevs/dracut/pull/900

Hence it would be good if dhcp-multi could address the wicked case too.

**Comment by tblume** [comment]
> ANJALI> If it is not too much trouble could you send me output of cat /proc/cmdline with the system booted (with ibft and hence multiple ip= options enabled)? If it's not easily available, then fine.

You won't see much on the command line, since most of the parameters are automatically added at initrd creation time.
Especially, there are no ip= lines.
However, here you go:

-->
# cat /proc/cmdline 
BOOT_IMAGE=/boot/vmlinuz-5.9.8-2-default root=UUID=3fdc3143-8e24-49f0-b499-d8a004a1c614 splash=verbose mitigations=auto console=tty0 console=ttyS0
--<

The log at initrd creation time shows:

-->
[...]
I: *** Store current command line parameters ***
I: Stored kernel commandline:
I: rd.driver.pre=btrfs
I: rd.driver.pre=scsi_dh_alua rd.driver.pre=scsi_dh_emc rd.driver.pre=scsi_dh_rdac rd.driver.pre=dm_multipath 
I: rd.iscsi.ibft=1 rd.iscsi.firmware=1
I:  root=/dev/mapper/360014057268ef2b5751436d84a0b3c0e-part2 rootfstype=btrfs rootflags=rw,relatime,space_cache,subvolid=256,subvol=/@,subvol=@
[...]
--<

**Comment by anjalidk** [comment]
@tblume

Got it. Looking at the info though, it looks like wicked will eventually 
be moved out of network-legacy in it's own module (35network-wicked)? So 
in that case, network-legacy won't have wicked anymore? If it's a 
temporary thing, then it's probably not worth the effort to port to wicked?

Also, parallelization may or may not work for wicked, depending on how 
parallel the wicked code itself is. Since dhclient can be run in 
parallel, hence the scope of this work was smaller. Also it needs access 
and information about SUSE which will need more time and effort. For 
making it work on wicked, we can create a new work item and address it 
separately, rather than clumping it with this work? I do think it would 
be good to eventually apply this to wicked as well. Let me know what you 
think?

For now, since single-dhcp code is independent of the ip=dhcp case, and 
wicked is applicable only to ip=dhcp, can we caveat this saying this is 
applicable to non-wicked/non-SUSE cases only? That is, it will be a 
no-op if ip=single-dhcp is added on SUSE/wicked. After wicked code is 
committed, I can add a check to see if the backend is wicked, then 
ignore the single-dhcp case (I cannot add that right now since code does 
not exist).

Anjali

**Comment by anjalidk** [comment]
How do I make changes to the dracut.cmdline manpage? Is it in the sources?

Anjali

**Comment by danimo** [comment]
> How do I make changes to the dracut.cmdline manpage? Is it in the sources?

Yes, you should edit `dracut.cmdline.7.asc` (AsciiDoc format), then call `make doc` to create/update  `dracut.cmdline.7`, `man -l dracut.cmdline.7` allows a preview without running `make install`.

**Comment by anjalidk** [comment]
Thank you!

**Comment by tblume** [comment]
> @tblume Got it. Looking at the info though, it looks like wicked will eventually be moved out of network-legacy in it's own module (35network-wicked)? So in that case, network-legacy won't have wicked anymore? If it's a temporary thing, then it's probably not worth the effort to port to wicked? Also, parallelization may or may not work for wicked, depending on how parallel the wicked code itself is. Since dhclient can be run in parallel, hence the scope of this work was smaller. Also it needs access and information about SUSE which will need more time and effort. For making it work on wicked, we can create a new work item and address it separately, rather than clumping it with this work? I do think it would be good to eventually apply this to wicked as well. Let me know what you think? For now, since single-dhcp code is independent of the ip=dhcp case, and wicked is applicable only to ip=dhcp, can we caveat this saying this is applicable to non-wicked/non-SUSE cases only? That is, it will be a no-op if ip=single-dhcp is added on SUSE/wicked. After wicked code is committed, I can add a check to see if the backend is wicked, then ignore the single-dhcp case (I cannot add that right now since code does not exist). Anjali

Ok, for me.
@danimo an objections to merge?

**Comment by tblume** [comment]
After discussion with danimo we came to the conclusion that there could be a race window if the machine is connected to multiple subnets and the dhcp servers respond at the same time.
Therefore I will give a dhclient setup a try in my testing environment.
Will report results ASAP.

**Comment by tblume** [comment]
> After discussion with danimo we came to the conclusion that there could be a race window if the machine is connected to multiple subnets and the dhcp servers respond at the same time.
> Therefore I will give a dhclient setup a try in my testing environment.
> Will report results ASAP.

Checked on a fedora 33 and found that NetworkManager is using an internal client there:

[  486.947741] kvm118 NetworkManager[605]: <info>  [1607096835.4940] dhcp-init: Using DHCP client 'internal'

Could you add to the manpage that it also won't apply to NetworkManager?

**Comment by anjalidk** [comment]
Sure! Thanks. I am working on some fixes to improve boot time for 
Network manager as well.

Were you able to check for race conditions? I had added some checks to 
avoid race condition when multiple devices return DHCP success at same 
time..

Anjali

**Comment by anjalidk** [comment]
@tblume added to manpage that it does not apply to NM.

**Comment by anjalidk** [comment]
@tblume are we ok to merge?

**Comment by tblume** [comment]
> @tblume are we ok to merge?

Sorry, for the delay.
I have some troubles with an RHEL on iSCSI root test system.
The manual setup fails boot (iBFT works).
Maybe we can speed things up if you send the test scenarios you have covered, so I can skip those tests.

**Comment by anjalidk** [comment]
Thanks! You mean it fails with ip=single-dhcp?

Please go ahead and do all tests, I would prefer that. Do let me know if 
anything fails. I have on Oracle Linux for remote iSCSI boot and various 
other regression tests we have here.

Anjali

**Comment by tblume** [comment]
> Thanks! You mean it fails with ip=single-dhcp? 

Yes, using multiple ip= parameters for iscsi multipath root fails booting on RHEL, even without your patch.
I need to fix that before I can do the verification.

> Please go ahead and do all tests, I would prefer that. Do let me know if anything fails. I have on Oracle Linux for remote iSCSI boot and various other regression tests we have here. Anjali

Ok, will do ASAP, but might need until the beginning of next year due to christmas vacation.

**Comment by anjalidk** [comment]
No problem at all!Thanks for doing the testing!Btw could we add some of the test cases you are testing to the regression tests covered?AnjaliOn Dec 17, 2020 11:58 PM, tblume <notifications@github.com> wrote:

Thanks! You mean it fails with ip=single-dhcp?

Yes, using multiple ip= parameters for iscsi multipath root fails booting on RHEL, even without your patch.
I need to fix that before I can do the verification.

Please go ahead and do all tests, I would prefer that. Do let me know if anything fails. I have on Oracle Linux for remote iSCSI boot and various other regression tests we have here. Anjali

Ok, will do ASAP, but might need until the beginning of next year due to christmas vacation.

**Comment by anjalidk** [comment]
Also I will see if I can setup RH here on my side and do some tests in addition to the ones you are doing.AnjaliOn Dec 18, 2020 10:39 AM, anjali.k.kulkarni@oracle.com wrote:No problem at all!Thanks for doing the testing!Btw could we add some of the test cases you are testing to the regression tests covered?AnjaliOn Dec 17, 2020 11:58 PM, tblume <notifications@github.com> wrote:

Thanks! You mean it fails with ip=single-dhcp?

Yes, using multiple ip= parameters for iscsi multipath root fails booting on RHEL, even without your patch.
I need to fix that before I can do the verification.

Please go ahead and do all tests, I would prefer that. Do let me know if anything fails. I have on Oracle Linux for remote iSCSI boot and various other regression tests we have here. Anjali

Ok, will do ASAP, but might need until the beginning of next year due to christmas vacation.

**Comment by tblume** [comment]
> Also I will see if I can setup RH here on my side and do some tests in addition to the ones you are doing.

Thanks!

AnjaliOn Dec 18, 2020 10:39 AM, anjali.k.kulkarni@oracle.com wrote:No problem at all!Thanks for doing the testing!Btw could we add some of the test cases you are testing to the regression tests covered?

I'm going to look into the regression tests and see what can be added.

**Comment by anjalidk** [comment]
@tblume, can you send me documentation or some pointers on how to build 
the checked out code for RH?

Anjali

**Comment by tblume** [comment]
> @tblume, can you send me documentation or some pointers on how to build the checked out code for RH? Anjali

@anjalidk I'm using open buildservice for building my testpackages.
It is fetching the dracut sources from my git repo and is building it for several systems.
You could use my project at:

https://build.opensuse.org/package/show/home:tsaupe:branches:Base:System:dracut-upstream-pr961/dracut

as example.
I still need to fix the package dependencies for building on RHEL, though.
Documentation about open buildservice is available at: 

https://openbuildservice.org/help/

**Comment by tblume** [comment]
Sorry, I'm afraid I have to give up trying to evaluate this on my multipathed iscsi root environment.
I just cannot get a setup that is using dhclient.
On SUSE dracut is using wicked, on Fedora it is using NetworkManager.
For RHEL 8 I'm lacking certain development libs to build the upstream package.
I then tried CentOS 8 stream but that doesn't really seem to work on my setup at all.
So, I'm at loss here. :(

Maybe someone from RH could review this instead?

**Comment by anjalidk** [comment]
Ok, so only multipath iscsi root fails or something else does? ie have 
any tests with this change tested/works for RH?

Btw, we have tested multipath iscsi root by our QA team here, on oracle 
linux, which is based on RH. These are the tests performed (see no. 5):


1. Network/Driver tests -  bnxt_en,ixgbe
2. KVM tests (libvirt, qemu smoke & post-smoke tests)
3. Test this on a system with Multiple NICs
4. DM MP SAN boot
5. iSCSI MP boot
6. Install & boot over X8-2,X7-2,X6-2,X5-8,X4-2,X3-2 models
7. iscsi functional tests
8. kdump test over local filesystem

What issue are you seeing in multipath iscsi root? Perhaps I can ask our 
QA to test again and check. From RH side, whom do you recommend could 
test this?

Also one more thing - for Fedora, would it be possible for you to 
disable Network manager and check?

Anjali

**Review by haraldh — APPROVED** [review]

**Comment by tblume** [comment]
> Ok, so only multipath iscsi root fails or something else does? ie have any tests with this change tested/works for RH? 

I only tested the case where multiple interfaces in the initrd would need to be setup up for booting.
Typically this is network root provided via multiple network interfaces.

> Btw, we have tested multipath iscsi root by our QA team here, on oracle linux, which is based on RH. These are the tests performed (see no. 5): 1. Network/Driver tests -  bnxt_en,ixgbe 2. KVM tests (libvirt, qemu smoke & post-smoke tests) 3. Test this on a system with Multiple NICs 4. DM MP SAN boot 5. iSCSI MP boot 6. Install & boot over X8-2,X7-2,X6-2,X5-8,X4-2,X3-2 models 7. iscsi functional tests 8. kdump test over local filesystem What issue are you seeing in multipath iscsi root? Perhaps I can ask our QA to test again and check. From RH side, whom do you recommend could test this? 

Seems that Harald did. :)

Also one more thing - for Fedora, would it be possible for you to disable Network manager and check? Anjali

Not sure how to disable NetworkManager in the initrd.
But since it is approved now, I guess we can skip that.

**Comment by anjalidk** [comment]
Thanks Harald and @tblume ! Do we need one more reviewer?

**Comment by johannbg** [comment]
@anjalidk you will need to rebase and fix your comment message to pass the commisery check. 
Change it to something like... 
docs: <your commit msg here for documentation changes, all lower case>
feat(network): <your commit msg here for the network module, all lower case>
feat(network-legacy): <your commit msg here for network-legacy module all lower case>

**Comment by anjalidk** [comment]
Thanks! I am looking at the commissionary errors..
________________________________

**Review by johannbg — APPROVED** [review]
You still need to fix *all* your commit messages to pass the commisery check but other than that looks good to me.

**Comment by anjalidk** [comment]
Thank you! I fixed the commissionary errors on branch parallel-dhcp, but it is still failing at the same place! Any idea why?
________________________________

**Comment by johannbg** [comment]
@anjalidk whatever you did seems to have trigger commisery to go through the entire git log from the looks of it and now is failing on old commits. That's a new one...

**Comment by danimo** [comment]
I think we can merge this now once you have resolved this conflicts. Please also make sure to rebase the changes, the merge commit makes this very hard to trace. Thank you.

**Comment by anjalidk** [comment]
Thank you - I will just create a new branch and pick all commits from parallel-dhcp and commit there - since rebasing has changed commit IDs of all commits until my commit - (I tried to rebase from top to my commit, but that is modifying commit id of all other commits after mine till TOT as well).
Anjali
________________________________

**Comment by anjalidk** [comment]
I created a new branch (parallel-dhcp-2) and a new pull request with all changes from parallel-dhcp branch, pull request is here:

https://github.com/dracutdevs/dracut/pull/1048

**Pull request closed without merging**
